File: cli/internal/agentauthz/powershell_command.go
package agentauthz

import (
	"regexp"
	"strings"
)

var powerShellMetricExecutablePattern = `(?:"[^"\r\n]*qdm-metric-cli(?:\.exe)?"|'[^'\r\n]*qdm-metric-cli(?:\.exe)?'|\$env:QDM_METRIC_CLI|(?:[A-Za-z]:)?(?:\.{0,2}[\\/])?(?:[^\s;|&'"` + "`" + `]+[\\/])*qdm-metric-cli(?:\.exe)?|qdm-metric-cli(?:\.exe)?)`

func IsPowerShellMetricAnalysisExecute(command string) bool {
	return powerShellInvocationRegexp(`analysis\s+execute`).MatchString(MaskPowerShellRegions(command))
}

func IsPowerShellMetricAuthDescribe(command string) bool {
	return powerShellInvocationRegexp(`auth\s+describe`).MatchString(MaskPowerShellRegions(command))
}

func IsPowerShellMetricAuthzGatedCommand(command string) bool {
	return powerShellInvocationRegexp(metricAuthzSubcommandPattern).MatchString(MaskPowerShellRegions(command))
}

func PowerShellMetricInvocationCount(command string) int {
	re := powerShellInvocationRegexp(metricAuthzSubcommandPattern)
	return len(re.FindAllStringIndex(MaskPowerShellRegions(command), -1))
}

func PowerShellCommandHasModelAuthFlags(command string) bool {
	skeleton := MaskPowerShellRegions(command)
	return regexp.MustCompile(`(?i)(?:^|\s)--(?:data-auth|auth-blob|auth-json|auth-user-id)\b`).MatchString(skeleton)
}

func powerShellInvocationRegexp(subcommand string) *regexp.Regexp {
	return regexp.MustCompile(`(?im)(?:^|[\r\n;|]|&&)\s*(?:\$[A-Za-z_][\w]*\s*=\s*)?(?:&\s*)?` + powerShellMetricExecutablePattern + `\s+` + subcommand + `\b`)
}

func RewritePowerShellMetricCLIInvocation(command, metricCLIPath string) string {
	if strings.TrimSpace(command) == "" || strings.TrimSpace(metricCLIPath) == "" {
		return command
	}
	return rewritePowerShellMetricCLIExecutable(command, PowerShellQuote(metricCLIPath))
}

func rewritePowerShellMetricCLIExecutable(command, replacement string) string {
	skeleton := MaskPowerShellRegions(command)
	re := regexp.MustCompile(`(?im)(?:^|[\r\n;|]|&&)(\s*)(?:\$[A-Za-z_][\w]*\s*=\s*)?(&\s*)?(` + powerShellMetricExecutablePattern + `)(\s+)(` + metricAuthzSubcommandPattern + `)\b`)
	matches := re.FindAllStringSubmatchIndex(skeleton, -1)
	if len(matches) == 0 {
		return command
	}
	var out strings.Builder
	last := 0
	for _, match := range matches {
		execStart, execEnd := match[6], match[7]
		ampStart := match[4]
		out.WriteString(command[last:execStart])
		if ampStart < 0 {
			out.WriteString("& ")
		}
		out.WriteString(replacement)
		last = execEnd
	}
	out.WriteString(command[last:])
	return out.String()
}

func StripPowerShellAuthFlags(command string) string {
	return stripAuthFlagsWithSkeleton(command, MaskPowerShellRegions(command))
}

func InjectPowerShellDataAuth(command, blob, metricCLIPath string) string {
	cleaned := StripPowerShellAuthFlags(command)
	cleaned = RewritePowerShellMetricCLIInvocation(cleaned, metricCLIPath)
	return insertPowerShellFlags(cleaned, " --data-auth --auth-blob "+PowerShellQuote(blob), "execute")
}

func InjectPowerShellAuthDescribeBlob(command, blob, metricCLIPath string) string {
	cleaned := StripPowerShellAuthFlags(command)
	cleaned = RewritePowerShellMetricCLIInvocation(cleaned, metricCLIPath)
	return insertPowerShellFlags(cleaned, " --auth-blob "+PowerShellQuote(blob), "describe")
}

func insertPowerShellFlags(command, flags, anchorWord string) string {
	skeleton := MaskPowerShellRegions(command)
	subcommand := metricAuthzSubcommandPattern
	if anchorWord == "describe" {
		subcommand = `auth\s+describe`
	}
	invocation := powerShellInvocationRegexp(subcommand).FindStringIndex(skeleton)
	if invocation == nil {
		return command
	}
	subcommandMatch := regexp.MustCompile(`(?i)` + subcommand).FindStringIndex(skeleton[invocation[0]:invocation[1]])
	if subcommandMatch == nil {
		return command
	}
	anchorEnd := invocation[0] + subcommandMatch[1]
	tail := skeleton[anchorEnd:]
	op := regexp.MustCompile(`\s(?:\||&&|;|2?>|1?>|&>)`).FindStringIndex(tail)
	if op == nil {
		return command + flags
	}
	position := anchorEnd + op[0]
	return command[:position] + flags + command[position:]
}

func PowerShellQuote(value string) string {
	return "'" + strings.ReplaceAll(value, "'", "''") + "'"
}

func MaskPowerShellRegions(command string) string {
	if command == "" {
		return ""
	}
	masked := []byte(command)
	space := func(from, to int) {
		for i := from; i < to && i < len(masked); i++ {
			if masked[i] != '\r' && masked[i] != '\n' {
				masked[i] = ' '
			}
		}
	}
	for i := 0; i < len(command); {
		if i+2 < len(command) && command[i] == '@' && (command[i+1] == '\'' || command[i+1] == '"') && (command[i+2] == '\r' || command[i+2] == '\n') {
			quote := command[i+1]
			endMarker := string([]byte{'\n', quote, '@'})
			end := strings.Index(command[i+2:], endMarker)
			if end < 0 {
				space(i, len(command))
				break
			}
			end += i + 2 + len(endMarker)
			space(i, end)
			i = end
			continue
		}
		if command[i] == '#' {
			end := strings.IndexByte(command[i:], '\n')
			if end < 0 {
				space(i, len(command))
				break
			}
			space(i, i+end)
			i += end
			continue
		}
		if command[i] == '\'' || command[i] == '"' {
			quote := command[i]
			j := i + 1
			for j < len(command) {
				if quote == '\'' && command[j] == '\'' && j+1 < len(command) && command[j+1] == '\'' {
					j += 2
					continue
				}
				if quote == '"' && command[j] == '`' && j+1 < len(command) {
					j += 2
					continue
				}
				if command[j] == quote {
					break
				}
				j++
			}
			if j >= len(command) {
				space(i+1, len(command))
				break
			}
			if !isPowerShellMetricExecutable(command[i+1 : j]) {
				space(i+1, j)
			}
			i = j + 1
			continue
		}
		i++
	}
	return string(masked)
}

func isPowerShellMetricExecutable(value string) bool {
	value = strings.TrimSpace(strings.ReplaceAll(value, "\\", "/"))
	lower := strings.ToLower(value)
	if lower == "$env:qdm_metric_cli" {
		return true
	}
	parts := strings.Split(lower, "/")
	base := parts[len(parts)-1]
	return base == "qdm-metric-cli" || base == "qdm-metric-cli.exe"
}

File: packages/data-harness-cli/src/lib/authz/powershell-command.js
import { stripAuthFlagsWithSkeleton } from "./metric-command.js";

const powerShellMetricExecutablePattern =
  String.raw`(?:"[^"\r\n]*qdm-metric-cli(?:\.exe)?"|'[^'\r\n]*qdm-metric-cli(?:\.exe)?'|\$env:QDM_METRIC_CLI|(?:[A-Za-z]:)?(?:\.{0,2}[\\/])?(?:[^\s;|&'` +
  "`" +
  String.raw`]+[\\/])*qdm-metric-cli(?:\.exe)?|qdm-metric-cli(?:\.exe)?)`;

export function isPowerShellMetricAnalysisExecute(command) {
  return powerShellInvocationRegexp(String.raw`analysis\s+execute`).test(maskPowerShellRegions(command));
}

export function isPowerShellMetricAuthDescribe(command) {
  return powerShellInvocationRegexp(String.raw`auth\s+describe`).test(maskPowerShellRegions(command));
}

export function isPowerShellMetricAuthzGatedCommand(command) {
  return isPowerShellMetricAnalysisExecute(command) || isPowerShellMetricAuthDescribe(command);
}

export function powerShellMetricInvocationCount(command) {
  const re = powerShellInvocationRegexp(String.raw`(?:analysis\s+execute|auth\s+describe)`);
  const skeleton = maskPowerShellRegions(command);
  return [...skeleton.matchAll(new RegExp(re.source, `${re.flags}g`))].length;
}

export function powerShellCommandHasModelAuthFlags(command) {
  return /(?:^|\s)--(?:data-auth|auth-blob|auth-json)\b/i.test(maskPowerShellRegions(command));
}

function powerShellInvocationRegexp(subcommand) {
  return new RegExp(
    `(?:^|[\\r\\n;|]|&&)\\s*(?:\\$[A-Za-z_][\\w]*\\s*=\\s*)?(?:&\\s*)?${powerShellMetricExecutablePattern}\\s+${subcommand}\\b`,
    "im",
  );
}

export function rewritePowerShellMetricCLIInvocation(command, metricCLIPath) {
  if (!String(command || "").trim() || !String(metricCLIPath || "").trim()) return command;
  return rewritePowerShellMetricCLIExecutable(command, powerShellQuote(metricCLIPath));
}

function rewritePowerShellMetricCLIExecutable(command, replacement) {
  const skeleton = maskPowerShellRegions(command);
  const re = new RegExp(
    `(?:^|[\\r\\n;|]|&&)(\\s*)(?:\\$[A-Za-z_][\\w]*\\s*=\\s*)?(&\\s*)?(${powerShellMetricExecutablePattern})(\\s+)(?:analysis\\s+execute|auth\\s+describe)\\b`,
    "gim",
  );
  const matches = [];
  let match;
  while ((match = re.exec(skeleton))) {
    matches.push(match);
  }
  if (matches.length === 0) return command;
  let out = "";
  let last = 0;
  for (const m of matches) {
    const execStart = m.index + m[0].indexOf(m[3]);
    const execEnd = execStart + m[3].length;
    const amp = m[2];
    out += command.slice(last, execStart);
    if (amp == null) out += "& ";
    out += replacement;
    last = execEnd;
  }
  return out + command.slice(last);
}

export function stripPowerShellAuthFlags(command) {
  return stripAuthFlagsWithSkeleton(command, maskPowerShellRegions(command));
}

export function injectPowerShellDataAuth(command, blob, metricCLIPath) {
  let cleaned = stripPowerShellAuthFlags(command);
  cleaned = rewritePowerShellMetricCLIInvocation(cleaned, metricCLIPath);
  return insertPowerShellFlags(cleaned, ` --data-auth --auth-blob ${powerShellQuote(blob)}`, "execute");
}

export function injectPowerShellAuthDescribeBlob(command, blob, metricCLIPath) {
  let cleaned = stripPowerShellAuthFlags(command);
  cleaned = rewritePowerShellMetricCLIInvocation(cleaned, metricCLIPath);
  return insertPowerShellFlags(cleaned, ` --auth-blob ${powerShellQuote(blob)}`, "describe");
}

function insertPowerShellFlags(command, flags, anchorWord) {
  const skeleton = maskPowerShellRegions(command);
  let subcommand = String.raw`analysis\s+execute`;
  if (anchorWord === "describe") subcommand = String.raw`auth\s+describe`;
  const invocation = powerShellInvocationRegexp(subcommand).exec(skeleton);
  if (!invocation) return command;
  const segment = skeleton.slice(invocation.index, invocation.index + invocation[0].length).toLowerCase();
  const relativeAnchor = segment.lastIndexOf(anchorWord.toLowerCase());
  if (relativeAnchor < 0) return command;
  const anchorEnd = invocation.index + relativeAnchor + anchorWord.length;
  const tail = skeleton.slice(anchorEnd);
  const op = /\s(?:\||&&|;|2?>|1?>|&>)/.exec(tail);
  if (!op) return command + flags;
  const position = anchorEnd + op.index;
  return command.slice(0, position) + flags + command.slice(position);
}

export function powerShellQuote(value) {
  return `'${String(value).replaceAll("'", "''")}'`;
}

export function maskPowerShellRegions(command) {
  if (!command) return "";
  const masked = Buffer.from(command, "utf8");
  const space = (from, to) => {
    for (let i = from; i < to && i < masked.length; i++) {
      if (masked[i] !== 0x0d && masked[i] !== 0x0a) masked[i] = 0x20;
    }
  };
  const text = command;
  for (let i = 0; i < text.length; ) {
    if (
      i + 2 < text.length &&
      text[i] === "@" &&
      (text[i + 1] === "'" || text[i + 1] === '"') &&
      (text[i + 2] === "\r" || text[i + 2] === "\n")
    ) {
      const quote = text[i + 1];
      const endMarker = `\n${quote}@`;
      const endRel = text.indexOf(endMarker, i + 2);
      if (endRel < 0) {
        space(i, text.length);
        break;
      }
      const end = endRel + endMarker.length;
      space(i, end);
      i = end;
      continue;
    }
    if (text[i] === "#") {
      const end = text.indexOf("\n", i);
      if (end < 0) {
        space(i, text.length);
        break;
      }
      space(i, end);
      i = end;
      continue;
    }
    if (text[i] === "'" || text[i] === '"') {
      const quote = text[i];
      let j = i + 1;
      while (j < text.length) {
        if (quote === "'" && text[j] === "'" && j + 1 < text.length && text[j + 1] === "'") {
          j += 2;
          continue;
        }
        if (quote === '"' && text[j] === "`" && j + 1 < text.length) {
          j += 2;
          continue;
        }
        if (text[j] === quote) break;
        j += 1;
      }
      if (j >= text.length) {
        space(i + 1, text.length);
        break;
      }
      if (!isPowerShellMetricExecutable(text.slice(i + 1, j))) {
        space(i + 1, j);
      }
      i = j + 1;
      continue;
    }
    i += 1;
  }
  return masked.toString("utf8");
}

function isPowerShellMetricExecutable(value) {
  value = value.trim().replaceAll("\\", "/");
  const lower = value.toLowerCase();
  if (lower === "$env:qdm_metric_cli") return true;
  const parts = lower.split("/");
  const base = parts[parts.length - 1];
  return base === "qdm-metric-cli" || base === "qdm-metric-cli.exe";
}
